### PR TITLE
Add OptionalTypeDesc to the AST

### LINKED ISSUE #867: Add OptionalTypeDesc to the AST
Currently we represent both T|() and T? similarly in the AST (Both as BinaryTypeDesc with right as BuiltinTypeDesc null) with only difference being for T? case BinaryTypeDesc’s right having same start and end pos. Instead we need to make this difference explicit in the AST.

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
Add OptionalTypeDesc to the AST
Resolves #867

## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: compiler/modules/front.syntax/ast.bal [hunk 1]
```diff
@@ -459,6 +459,12 @@ public type GroupingTypeDesc record {|
     TypeDesc innerTd;
 |};
 
+public type OptionalTypeDesc record {|
```
Comment (jclark) [inline compiler/modules/front.syntax/ast.bal]: OptionalTypeDesc (and GroupingTypeDesc) are not consistent with UnaryTypeDesc below.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/ast.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/ast.bal]: fixed

Comment (jclark) [reply compiler/modules/front.syntax/ast.bal]: The field names need to be distinct.

This still isn't consistent: UnaryTypeDesc has an `op`.

For any change to the AST please propose a design in the issue before implementing.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/ast.bal (hunk 1 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/ast.bal]: fixed

File: compiler/modules/front.syntax/ast.bal [hunk 2]
```diff
@@ -524,9 +518,22 @@ public type BinaryTypeDesc record {|
...
+public type PrefixTypeDesc record {|
     *PositionFields;
-    UnaryTypeOp op;
+    PrefixTypeOp op;
+    TypeDesc td;
+|};
+
+public const PostfixTypeOp = "?";
+
+public type PostfixTypeDesc record {|
```
Comment (jclark) [inline compiler/modules/front.syntax/ast.bal]: Remember that type names are just aliases.  So what you have defined is completely equivalent to having a UnaryTypeDesc record with an `op` field of `"!"|"?"|"("`. That is OK as a design but you can be written much more simply than what you have here.

**Review by jclark — CHANGES_REQUESTED** [review]

File: compiler/modules/front.syntax/ast.bal (hunk 2 above)
Comment (heshanpadmasiri) [reply compiler/modules/front.syntax/ast.bal]: fixed

**Pull request merged**